coderd/workspaceapps/stats.go

```diff
@@ -0,0 +1,378 @@
+	err := r.db.InTx(func(tx database.Store) error {
+		var batch database.InsertWorkspaceAppStatsParams
+		for _, stat := range stats {
+			batch.UserID = append(batch.UserID, stat.UserID)
+			batch.WorkspaceID = append(batch.WorkspaceID, stat.WorkspaceID)
+			batch.AgentID = append(batch.AgentID, stat.AgentID)
+			batch.AccessMethod = append(batch.AccessMethod, string(stat.AccessMethod))
+			batch.SlugOrPort = append(batch.SlugOrPort, stat.SlugOrPort)
+			batch.SessionID = append(batch.SessionID, stat.SessionID)
+			batch.SessionStartedAt = append(batch.SessionStartedAt, stat.SessionStartedAt)
+			batch.SessionEndedAt = append(batch.SessionEndedAt, stat.SessionEndedAt)
+			batch.Requests = append(batch.Requests, int32(stat.Requests))
+
+			if len(batch.UserID) >= r.batchSize {
+				err := tx.InsertWorkspaceAppStats(ctx, batch)
+				if err != nil {
+					return err
+				}
+
+				// Reset batch.
+				batch = database.InsertWorkspaceAppStatsParams{}
+			}
+		}
+		if len(batch.UserID) > 0 {
+			err := tx.InsertWorkspaceAppStats(ctx, batch)
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}, nil)
+	if err != nil {
+		return xerrors.Errorf("insert workspace app stats failed: %w", err)
+	}
+
+	return nil
+}
+
+// This should match the database unique constraint.
+type statsGroupKey struct {
+	StartTimeTrunc time.Time
+	UserID         uuid.UUID
+	WorkspaceID    uuid.UUID
+	AgentID        uuid.UUID
+	AccessMethod   AccessMethod
+	SlugOrPort     string
+}
+
+func (s StatsReport) groupKey(windowSize time.Duration) statsGroupKey {
+	return statsGroupKey{
+		StartTimeTrunc: s.SessionStartedAt.Truncate(windowSize),
+		UserID:         s.UserID,
+		WorkspaceID:    s.WorkspaceID,
+		AgentID:        s.AgentID,
+		AccessMethod:   s.AccessMethod,
+		SlugOrPort:     s.SlugOrPort,
+	}
+}
+
+// StatsCollector collects workspace app StatsReports and reports them
+// in batches, stats compaction is performed for short-lived sessions.
+type StatsCollector struct {
+	opts StatsCollectorOptions
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	done   chan struct{}
+
+	mu               sync.Mutex                       // Protects following.
+	statsBySessionID map[uuid.UUID]*StatsReport       // Track unique sessions.
+	groupedStats     map[statsGroupKey][]*StatsReport // Rolled up stats for sessions in close proximity.
+	backlog          []StatsReport                    // Stats that have not been reported yet (due to error).
+}
+
+type StatsCollectorOptions struct {
+	Logger   *slog.Logger
+	Reporter StatsReporter
+	// ReportInterval is the interval at which stats are reported, both partial
+	// and fully formed stats.
+	ReportInterval time.Duration
+	// RollupWindow is the window size for rolling up stats, session shorter
+	// than this will be rolled up and longer than this will be tracked
+	// individually.
+	RollupWindow time.Duration
+
+	// Options for tests.
+	Flush <-chan chan<- struct{}
+	Now   func() time.Time
```

Comment:
Suggestion: allow setting a `chan time.Time` for use in unit tests. This will allow more fine-grained control over time.

Reply:
I think the now func is more flexible than a chan. We can even implement the chan with it, if we want: `Now: func() time.Time { <-testTimeCh }`.

Reply:
Fair :+1: